src/server/coprocessor/dag.rs

```diff
@@ -0,0 +1,145 @@
+               eval_ctx: Rc<EvalContext>)
+               -> DAGContext<'s> {
+        DAGContext {
+            req: req,
+            columns: vec![],
+            ranges: ranges,
+            snap: snap,
+            has_aggr: false,
+            eval_ctx: eval_ctx,
+            chunks: vec![],
+        }
+    }
+
+    pub fn validate_dag(&mut self) -> Result<()> {
+        let execs = self.req.get_executors();
+        if execs.is_empty() {
+            return Err(box_err!("dag copr has not executor"));
+        }
+        // check whether first exec is *scan and get the column info
+        let first = &execs[0];
+        match first.get_tp() {
+            ExecType::TypeTableScan => {
+                self.columns = first.get_tbl_scan().get_columns().to_vec();
+            }
+            ExecType::TypeIndexScan => {
+                self.columns = first.get_idx_scan().get_columns().to_vec();
+            }
+            _ => {
+                return Err(box_err!("first exec type should be *Scan, but get {:?}",
+                                    first.get_tp()))
+            }
+        }
+        // check whether dag has a aggregation action and take a flag
+        if execs.iter().rfind(|&exec| exec.get_tp() == ExecType::TypeAggregation).is_some() {
+            self.has_aggr = true;
+        }
+        Ok(())
+    }
+
+    // FIXME seperate first exec build action from `build_dag`
+    // since it will generte some mutable conflict when putting together
+    pub fn build_first(&'s self,
+                       mut first: Executor,
+                       statistics: &'s mut Statistics)
+                       -> Box<DAGExecutor + 's> {
+        match first.get_tp() {
+            ExecType::TypeTableScan => {
+                Box::new(TableScanExecutor::new(first.take_tbl_scan(),
+                                                self.ranges.clone(),
+                                                self.snap,
+                                                statistics,
+                                                self.req.get_start_ts(),
+                                                IsolationLevel::SI))
+            }
+            ExecType::TypeIndexScan => {
+                Box::new(IndexScanExecutor::new(first.take_idx_scan(),
+                                                self.ranges.clone(),
+                                                self.snap,
+                                                statistics,
+                                                self.req.get_start_ts(),
+                                                IsolationLevel::SI))
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    pub fn build_dag(&'s self, statistics: &'s mut Statistics) -> Result<Box<DAGExecutor + 's>> {
+        let mut execs = self.req.get_executors().to_vec().into_iter();
+        let mut src = self.build_first(execs.next().unwrap(), statistics);
+        for mut exec in execs {
```

Comment:
yes, i had a try, but it makes code a little difficult to understand based on the origin logic which is already a little complicated.